### ISSUE TITLE
fail to collect references in release build

### ISSUE THREAD

**Issue by mgood7123** [Ravenbrook/mps#259]
jump to https://github.com/Ravenbrook/mps/issues/259#issuecomment-1729139318

i am unsure how to go about keeping roots alive as they get eligible for garbage collection immediately after creation

```cpp
    ManagedObjState state;

    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);
    
    managed_obj_t p = managed_obj_make(&state, malloc(500));
    printf("p is %p\n", p);

    // managed_obj_describe(&state);

    printf("first collect, p should not be collected\n");
    managed_obj_collect(&state); // should not collect p

    printf("if p has been collected then we have not managed to track it correctly\n");
    p = (managed_obj_t) 0x6; // no longer referenced

    printf("second collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("if p has NOT been collected then we have not managed to track it correctly\n");

    managed_obj_deinit(&state);
```
(i assume there would be more that needs to be done in my implementation (below) to correctly support the above tho i am unsure as to what would be needed)
```cpp
first collect, p should not be collected

object 0x7ffbf7100000 with pointer 0x12ea080 is being freed.
Total Collection duration: 0 seconds, 0 milliseconds, 192 microseconds

if p has been collected then we have not managed to track it correctly

second collect, p should be collected

Total Collection duration: 0 seconds, 0 milliseconds, 70 microseconds

if p has NOT been collected then we have not managed to track it correctly

deinit
deinit - collect
Total Collection duration: 0 seconds, 0 milliseconds, 43 microseconds
deinit - shutdown
```

the complete .h and .c file is https://gist.github.com/mgood7123/8a25971c90faa61397701361a0dd8be5

**Comment by mgood7123** [comment]
is there a way we can make this more scalable to larger amounts of allocations?

**Comment by mgood7123** [comment]
how feasible would it be to just register and destroy gc root structures as they are passed around? (eg, via C++ constructor/destructor semantics)

**Comment by rptb1** [comment]
Hello!

We've look at your code, and it has a problem with the registration of the stack root.  Your are registering the stack from within a function, and using *that* function's stack frame as the [cold end](https://memory-pool-system.readthedocs.io/en/version-1.118/glossary/c.html#term-cold-end) of the stack. See https://gist.github.com/mgood7123/8a25971c90faa61397701361a0dd8be5#file-managed_object_obj-c-L434

This means that none of the functions calling managed_obj_make will have their stack frames included in the root, and objects they point to might be GC'd by the MPS.

The documentation section [Thread Roots](https://memory-pool-system.readthedocs.io/en/version-1.118/topic/root.html#thread-roots) describes how the stack root must be created.  All functions that have roots in registers must be called *by* the function that locates the cold end in the way shown.

In general, you do not need to create and destroy roots very often -- certainly not once per object.  The MPS is designed so that you register roots, then store pointers to objects you want to keep alive in those roots.  The typical case is that the roots are the stack and registers of your thread, and all objects (recursively) reachable from there are kept alive.  Preserving objects is just a matter of storing pointers to memory.

You definitely could have a kind of C++ "root object" that had a call to mps_root_create in its constructor and mps_root_destroy in its destructor, but that's the sort of thing we expect you do rarely: typically only once per process run.

We can't quite see from your code what you're trying to achieve.  Feel free to explain your application.  We're curious and would like to help.

Also we recommend you study the [Scheme Example](https://github.com/Ravenbrook/mps/tree/master/example/scheme) to see how it creates roots.

Thanks.

**Comment by mgood7123** [comment]
i tried using the stack bottom (GC_get_stack_base from `boehm`) as the thread stack root

which works correctly in debug build

but fails in a release build (no GC takes place)

**Comment by mgood7123** [comment]
> Hello!
> 
> We've look at your code, and it has a problem with the registration of the stack root. Your are registering the stack from within a function, and using _that_ function's stack frame as the [cold end](https://memory-pool-system.readthedocs.io/en/version-1.118/glossary/c.html#term-cold-end) of the stack. See https://gist.github.com/mgood7123/8a25971c90faa61397701361a0dd8be5#file-managed_object_obj-c-L434
> 
> This means that none of the functions calling managed_obj_make will have their stack frames included in the root, and objects they point to might be GC'd by the MPS.
> 
> The documentation section [Thread Roots](https://memory-pool-system.readthedocs.io/en/version-1.118/topic/root.html#thread-roots) describes how the stack root must be created. All functions that have roots in registers must be called _by_ the function that locates the cold end in the way shown.
> 
> In general, you do not need to create and destroy roots very often -- certainly not once per object. The MPS is designed so that you register roots, then store pointers to objects you want to keep alive in those roots. The typical case is that the roots are the stack and registers of your thread, and all objects (recursively) reachable from there are kept alive. Preserving objects is just a matter of storing pointers to memory.
> 
> You definitely could have a kind of C++ "root object" that had a call to mps_root_create in its constructor and mps_root_destroy in its destructor, but that's the sort of thing we expect you do rarely: typically only once per process run.
> 
> We can't quite see from your code what you're trying to achieve. Feel free to explain your application. We're curious and would like to help.
> 
> Also we recommend you study the [Scheme Example](https://github.com/Ravenbrook/mps/tree/master/example/scheme) to see [how it creates roots](https://github.com/Ravenbrook/mps/blob/554513b16461ec0bbf82856dfe6b0528518c3658/example/scheme/scheme.c#L4469-L4491).
> 
> There might be a misunderstanding of what we call a "root". See https://memory-pool-system.readthedocs.io/en/version-1.118/glossary/r.html#term-root . A root is not GC'd by the MPS. A root is a (set of) pointers _to_ objects managed by the MPS. So your original question "how to keep a root alive" might be trying to ask "how to keep an object referenced by a root alive"? The answer is that the MPS preserves all objects referenced by roots. If an object is dying, it is not referenced by a root. That could be because of the cold end problem I mentioned above.
> 
> Thanks.

ultimately i would like to create an api with MPS that allows me to basically have a stateful GC object that can "attach" and "detach" from threads

for example
```
// data is shared between T1 T2 and T3
//
mps_gc_1 for threads 1 2 3

// data is shared between T2 T4 and T5
//
// NOTE: mps_gc_1 will not touch data allocated by mps_gc_2 even though both are attached to thread 2
//
mps_gc_2 for threads 4 2 5

// dedicated MPS GC for thread 7, data is known not to be shared between any other thread
//
//   it is possible to localize data by carefully attaching and detaching from threads as data moves between threads
//
mps_gc_3 for thread 7

mps_gc 1 and 2 can co-exist with mps_gc 3 due to them running on seperate threads, unlike a global gc which runs on all threads even if data only exists on a single thread
```

**Comment by mgood7123** [comment]
i tried

```cpp
    ManagedObjState state;

    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);

```
```cpp
void managed_obj_init(ManagedObjState * state, size_t arena_size) {
    if (state == NULL) {
        managed_obj_error("state cannot be null");
    }
    memset(state, 0, sizeof(ManagedObjState));

    mps_res_t res;

// ...

/*

        A risk of using tagged pointers in registers and on the stack
        is that in some circumstances, an optimizing compiler might
        optimize away the tagged pointer, keeping only the untagged
        version of the pointer. In this situation the pointer would be
        ignored and if it was the last reference to the object the MPS
        might incorrectly determine that it was dead. 

        #. Choose to tag pointers with zero, setting ``scan_area`` to
           :c:func:`mps_scan_area_tagged` and setting ``pattern`` to
           zero.

*/

    void *marker = state;

    res = mps_root_create_thread_tagged(
        &state->thread_stack_root,
        state->arena,
        mps_rank_ambig(),
        (mps_rm_t)0,
        state->thread,
        mps_scan_area_tagged_or_zero,
        (mps_word_t)0,
        (mps_word_t)0,
        marker
    );

    if (res != MPS_RES_OK) printf("Couldn't create root");
}
```

but release build still fails to find the pointers where a debug build would successfully find the pointers

**Comment by mgood7123** [comment]
if we build using `gcc` and `g++` instead of `clang` and `clang++` then a release build successfully finds our pointers and GC's them

**Comment by mgood7123** [comment]
if i make my pointer `volatile` then it gets collected at `-O2` and `-O3` but i cannot force other libraries to use volatile

```c
    volatile managed_obj_t p1 = managed_obj_make(&state, malloc(500));
    volatile managed_obj_t p2 = managed_obj_make(&state, malloc(500));

// ...

    p1 = (managed_obj_t) 0x6; // no longer referenced
    p2 = (managed_obj_t) 0x6; // no longer referenced

// ...

    printf("second collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

```

**Comment by mgood7123** [comment]
this seems like with optimizations, attempting to bring a variable into the local stack of a function MAY bind it to the frame, resulting in it being uncollectable until the frame ends

this in turn CAN be troublesom if we attempt to init and destroy the GC in the same frame, since the GC cannot reclaim a variable that is bound to the same frame as it (or even may leak to a parent frame variable (would technically be UB as it would yeild a phantom reference i think? or similar to one, as the reference exists but is untracked since the GC has been destroyed and the reference was still alive at such point)

**Comment by mgood7123** [comment]
one solution would be to collect all references upon the arena being destroyed (regardless of weather such are reachable or not)

but i do not know how to do this

**Comment by mgood7123** [comment]
it seems we can do so via

```c
static Res rootDestroy(Root root, void *p)
{
  /* we are inside a locked arena, we cannot */
  /* use 'mps_root_destroy' since it locks the arena */

  AVERT(Root, root);
  
  Arena arena;

  arena = RootArena(root);

  RootDestroy(root);

  return ResOK;
}

struct mps_thr_s;

// all mps ThreadStruct contain this layout

typedef struct mps_thr_s {      /* ANSI fake thread structure */
  Sig sig;                      /* design.mps.sig.field */
  Serial serial;                /* from arena->threadSerial */
  Arena arena;                  /* owning arena */
  RingStruct arenaRing;         /* attaches to arena */
  /* extra data per impl */
} ThreadStruct;

void destroy_roots(mps_arena_t mps_arena) {
  Arena arena = (Arena)mps_arena;

  /* lock the arena to prevent any roots from */
  /* being created while we are iterating them */

  ArenaEnter(arena);

  AVER(ArenaGlobals(arena)->clamped);          /* .assume.parked */
  AVER(arena->busyTraces == TraceSetEMPTY);    /* .assume.parked */

  Globals arenaGlobals = ArenaGlobals(arena);

  AVERT(Globals, arenaGlobals);

  RootsIterate(arenaGlobals, rootDestroy, 0);
  
  /* unregister all registered threads since */
  /* we have no roots associated with them */

  /*
  _`.req.register.multi`: It must be possible to register the same
  thread multiple times. (This is needed to support the situation where
  a program that does not use the MPS is calling into MPS-using code
  from multiple threads. On entry to the MPS-using code, the thread can
  be registered, but it may not be possible to ensure that the thread is
  deregistered on exit, because control may be transferred by some
  non-local mechanism such as an exception or ``longjmp()``. We don't
  want to insist that the client program keep a table of threads it has
  registered, because maintaining the table might require allocation,
  which might provoke a collection. See request.dylan.160252_.)


  *  Register/Deregister
  *
  *  Explicitly register/deregister a thread on the arena threadRing.
  *  Register returns a "Thread" value which needs to be used
  *  for deregistration.
  *
  *  Threads must not be multiply registered in the same arena.

  */

  Ring threadRing = ArenaThreadRing(arena);
  Ring node, next;
  AVERT(Ring, threadRing);
  RING_FOR(node, threadRing, next) {
    Thread thread = RING_ELT(Thread, arenaRing, node);

    AVER(ThreadCheckSimple(thread));
    
    AVER(arena == ThreadArena(thread)); /* is the thread arena always us? */

    ThreadDeregister(thread, arena);
  }

  ArenaLeave(arena);
}
```

and then collect the arena

```cpp
    mps_arena_park(state->arena);        /* ensure no collection is running */

    printf("deinit - collect\n");

    destroy_roots(state->arena); // destroy all roots and unregister all threads from the given arena

    managed_obj_collect(state); // mps_arena_collect + mps_chat

    printf("deinit - collected\n");

    printf("deinit - shutdown\n");

    mps_ap_destroy(state->ap);       /* destroy ap before pool */
    mps_pool_destroy(state->pool);   /* destroy pool before fmt */
    mps_fmt_destroy(state->fmt);     /* destroy fmt before chain */
    mps_chain_destroy(state->chain); /* destroy chain before arena */
    mps_arena_destroy(state->arena);     /* last of all */
```

**Comment by mgood7123** [comment]
in a release build we get this

```cpp
    ManagedObjState state;

    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);
    
    managed_obj_t p1 = managed_obj_make(&state, malloc(500));
    managed_obj_t p2 = managed_obj_make(&state, malloc(500));

    p1 = (managed_obj_t) 0x6; // no longer referenced
    p2 = (managed_obj_t) 0x6; // no longer referenced

    managed_obj_deinit(&state);
```

```cpp
[ RUN      ] MPS_collect2.1
reserved and comitted object 0x7f98a3580000 with pointer 0x6991c0
reserved and comitted object 0x7f98a3580018 with pointer 0x6993c0
deinit - collect
object 0x7f98a3600000 with pointer 0x6991c0 is being freed.
object 0x7f98a3600018 with pointer 0x6993c0 is being freed.
Total Collection duration: 0 seconds, 0 milliseconds, 327 microseconds
deinit - collected
deinit - shutdown
[       OK ] MPS_collect2.1 (1 ms)
```

**Comment by mgood7123** [comment]
do note that this is intended to forcefully collect all references held by an arena

`provided no other arena also holds such` (not sure how to detect)

**Comment by rptb1** [comment]
Hello again!  We are very interested in what you are doing, but our resource has been limited for the past few days.  We *will* catch up.

I'll just point out that there is *no* communication between GCs in separate arenas.  Each arena is like a separate world.  Each client program (with its own object graph) is intended to have a single arena.  The use case for multiple arenas would be if there were, e.g. two separate language runtimes sharing a process, both linking to the MPS as a library, and calling each other via some sort of FFI mechanism.

Pools *within* an arena are fully co-operative.  The object graph can span pools and this will be handled correctly.

**Comment by mgood7123** [comment]
> Hello again! We are very interested in what you are doing, but our resource has been limited for the past few days. We _will_ catch up.
> 
> I'll just point out that there is _no_ communication between GCs in separate arenas. Each arena is like a separate world. Each client program (with its own object graph) is intended to have a single arena. The use case for multiple arenas would be if there were, e.g. two separate language runtimes sharing a process, both linking to the MPS as a library, and calling each other via some sort of FFI mechanism.
> 
> Pools _within_ an arena are fully co-operative. The object graph can span pools and this will be handled correctly.

so arena's `cannot` contain strong references to references allocated by other arena's ?

**Comment by mgood7123** [comment]
> do note
> 
> this is intended to forcefully collect all references held by an arena, with the intention of destroying the arena afterwards
> 
> `provided no other arena also holds such references` (not sure how to detect this)
> 
> this means that any references `currently in use` will become invalid (`unless another arena also has a reference (not sure how to detect this)`)
> 
> (similar to freeing a memory arena and all pointers to such memory becoming UB (use-after-free) )
> 
> these should be detectable with debuggers and sanitizers

incedentally this MAY be an alternative to `The only reliable way to ensure that all finalizable objects are finalized is to maintain a table of [weak references (1)](https://memory-pool-system.readthedocs.io/en/latest/glossary/w.html#term-weak-reference-1) to all such objects`

**Comment by rptb1** [comment]
> so arena's `cannot` contain `strong references` to references allocated by other arena's ?

Arenas can contain anything, but the MPS GC will not *automatically* find references (pointers) from one arena to another.  There is nothing to stop you creating roots in one arena for another arena, just like you would create roots from any other memory.  The arenas do not see each other: they are separate worlds.

> eg, a reference `cannot be owned by multiple arena's` ?

Please explain what you mean by "owned" here.

> assuming this to be the case, if an arena is being destroyed, how might we go about `migrating` `strong references` from one arena to another arena `(if needed)` ?

Please explain what you mean by "migrate" here.  You can copy memory from one arena to another, e.g. with `memcpy`.

When dealing with GC finalization, it's very important to realise that there is no guarantee of promptness. See [Cautions](https://memory-pool-system.readthedocs.io/en/latest/topic/finalization.html#cautions).  Finalization by the GC is provided as an optimisation, and should not be used as a necessary part of an algorithm.

> incedentally this MAY be an alternative to The only reliable way to ensure that all finalizable objects are finalized is to maintain a table of [weak references (1)](https://memory-pool-system.readthedocs.io/en/latest/glossary/w.html#term-weak-reference-1) to all such objects

This is an interesting point.  We don't currently guarantee that deleting all roots and collecting the world will recycle (and therefore finalize) every object, but we might be able to make that promise.  We'll consider it.

Also, internally, finalization is managed by a "guardian pool" that has a list of all remaining finalizable objects.  It may be possible to provide an interface to that list, so that a program can examine them all.

**Comment by rptb1** [comment]
> this is intended to forcefully collect all references held by an arena, with the intention of destroying the arena afterwards

There may be some misunderstanding between us here.  The MPS GC does not collect [references](https://memory-pool-system.readthedocs.io/en/latest/glossary/r.html#term-reference) (pointers to GC'd objects).  It collects [objects](https://memory-pool-system.readthedocs.io/en/latest/glossary/o.html#term-object) (blocks of memory).  Objects can *contain* references, discovered by [scanning](https://memory-pool-system.readthedocs.io/en/latest/glossary/s.html#term-scan-method).

**Comment by mgood7123** [comment]
> > eg, a reference `cannot be owned by multiple arena's` ?
> 
> Please explain what you mean by "owned" here.

never mind, i realized that arena's give out references, and they cannot "take ownership" of references

eg `ref r; arena_own(&r, arena); // doesnt work that way`

> Please explain what you mean by "migrate" here

say we have `ref_1` in `arena_1`, and we want to destroy `arena_1` while still keeping `ref_1` alive, by moving it from `arena_1` to `arena_2`, how might we do this?

> There may be some misunderstanding between us here. The MPS GC does not collect [references](https://memory-pool-system.readthedocs.io/en/latest/glossary/r.html#term-reference) (pointers to GC'd [objects](https://memory-pool-system.readthedocs.io/en/latest/glossary/o.html#term-object)). It collects objects (blocks of memory). Objects can contain references, discovered by [scanning](https://memory-pool-system.readthedocs.io/en/latest/glossary/s.html#term-scan-method).

yea, assuming we have `ref_40` and we set it to NULL, if upon collection it determines `ref_40` cannot be located it will finalize (if registered) and collect it

by deleting the roots themselves we basically force `ref_40` to be unlocatable even if it would have been locatable if the roots where still present

> finalization is managed by a "guardian pool" that has a list of all remaining finalizable objects. It may be possible to provide an interface to that list, so that a program can examine them all.

true, however our goal is to `explicitly finalize all unfinalized objects`, eg `turn all black objects to white objects, and keep all white objects as white objects` such that `everything in the arena is now white`

**Comment by rptb1** [comment]
> say we have `ref_1` in `arena_1`, and we want to destroy `arena_1` while still keeping `ref_1` alive, by moving it from `arena_1` to `arena_2`, how might we do this?

Are you saying ref_1 is a pointer inside an object allocated in arena_1?

The MPS will not examine references in arena_1 when it GCs arena_2.  Arenas do not communicate.  They are independent, separate worlds.  So ref_1 is *not ever* keeping anything in arena_2 alive.

The only way that arena_2 will know about ref_1 (which is outside arena_2) is if you create a root in arena_2 that includes ref_1.

From your questions here, and in #260 , you seem to be focussed on the management of external resources (file handles, etc.), not on the management of memory blocks.  Is that correct?  Are you trying to use the MPS mainly to manage external resources automatically?

**Comment by rptb1** [comment]
I'll also just repeat that the MPS is designed to have *one* arena per program.  The only use case for multiple arenas is if you have linked two independent programs (e.g. written by different organizations) into the same process, and they are sharing address space, but working independently.

**Comment by rptb1** [comment]
You are however welcome (and intended) to create multiple *pools* within your arena.  That might be what you need.  Let's say you create two instances of the AMC pool class, amcA and amcB.  Now, if an object in amc1 (objA) is keeping an object in amc2 (objB) alive, and you destroy pool amc1, then objB will (eventually) be recycled (and possibly finalized).  If you don't want that to happen, you need a pointer to objB in an object that is still alive, or in a root.  How you do that is up to you, but the simplest case is that you just do `objC.b = objB` where objC is some other object, which is presumably not dead because you're talking about it.

**Comment by mgood7123** [comment]
> This is an interesting point. We don't currently guarantee that deleting all roots and collecting the world will recycle (and therefore finalize) every object, but we might be able to make that promise. We'll consider it.

additionally, when since all roots are destroyed, reference resurrection should become effectively impossible

the `only` exception to this, is a root `can be manually re-registered` in the arena somehow, which requires a reference gaining access to the arena it belongs to, and then creating a root and then storing itself inside that root

**Comment by rptb1** [comment]
> true, however our goal is to `explicitly finalize all unfinalized objects`, eg `turn all black objects to white objects, and keep all white objects as white objects` such that `everything in the arena is now white`

See item 4 of finalization [Cautions](https://memory-pool-system.readthedocs.io/en/latest/topic/finalization.html#cautions):

> The only reliable way to ensure that all finalizable objects are finalized is to maintain a table of [weak references (1)](https://memory-pool-system.readthedocs.io/en/latest/glossary/w.html#term-weak-reference-1) to all such objects. The weak references don’t prevent the objects from being finalized, but you can iterate over the table at an appropriate point and finalize any remaining objects yourself.

We *might* be able to give access to the list of remaining finalized objects, but we'd have to think hard about whether we can do this safely.

**Comment by rptb1** [comment]
> additionally, since all roots are destroyed, reference resurrection should become effectively impossible

That's not quite true in the MPS.  When the MPS discovers a finalizable object, it add it to a [message queue in the arena](https://memory-pool-system.readthedocs.io/en/latest/topic/finalization.html#finalization-messages).  The message queue is not GCd, and so it is a bit like a root.  So the message itself resurrects the object until it is deleted.

In most scenarios where the MPS is part of a language run-time system, the program cannot run with all roots destroyed, because it cannot allocate any objects that won't die immediately!

However, if you're writing a program where only *some* objects are managed by the MPS, it can make sense.

> the `only` exception to this, is a root `can be manually re-registered` in the arena somehow, which requires a reference gaining access to the arena it belongs to, and then creating a root and then storing itself inside that root

This is not necessary, because the finalization message keeps the object alive (see above).

**Comment by mgood7123** [comment]
> > say we have `ref_1` in `arena_1`, and we want to destroy `arena_1` while still keeping `ref_1` alive, by moving it from `arena_1` to `arena_2`, how might we do this?
> 
> Are you saying ref_1 is a pointer inside an object allocated in arena_1?
> 
> The MPS will not examine references in arena_1 when it GCs arena_2. Arenas do not communicate. They are independent, separate worlds. So ref_1 is _not ever_ keeping anything in arena_2 alive.
> 
> The only way that arena_2 will know about ref_1 (which is outside arena_2) is if you create a root in arena_2 that includes ref_1.
> 
> From your questions here, and in #260 , you seem to be focussed on the management of external resources (file handles, etc.), not on the management of memory blocks. Is that correct? Are you trying to use the MPS mainly to manage external resources automatically?

like

```cpp
managed_obj_t managed_obj_make_with_finalizer(ManagedObjState * state, void * pointer, managed_obj_finalization_callback_t finalization_callback)
{
  managed_obj_t obj;
  size_t size = MANAGED_OBJECT_ALIGN_OBJ(sizeof(managed_obj_pointer_s));
  do {
    mps_res_t res = mps_reserve((mps_addr_t*)&obj, state->ap, size);
    if (res != MPS_RES_OK) managed_obj_error("out of memory in make_pointer");
    obj->pointer.type = MANAGED_OBJECT_TYPE_POINTER;
    obj->pointer.pointer = pointer;
    obj->pointer.finalization_callback = finalization_callback;
  } while(!mps_commit(state->ap, obj, size));
  printf("reserved and comitted object %p with pointer %p\n", obj, obj->pointer.pointer);

  state->total += sizeof(managed_obj_pointer_s);

  mps_finalize(state->arena, (mps_addr_t*)&obj);

  return obj;
}
```

```cpp

GC_CREATE (GC_1)

managed_obj_t ref_1 = managed_obj_make(GC_1);

// ...

// we now wish to migrate `ref_1` to a different arena, probably because this GC about to shut down and we want to retain `ref_`

managed_obj_t ref_1_copy = managed_obj_make(GC_2);

managed_object_migrate(ref_1, ref_1_copy); // migrate from GC_1 to GC_2

GC_SHUTDOWN(GC_1);

**Comment by rptb1** [comment]
Aha!  What you are calling `ref_1` is what we call an *object*, not a reference.  What we could call a reference is the field obj->pointer.pointer.

You can copy objects wherever you like, but be warned:

1. if your `ref_1->pointer.pointer` points to an object obj_2 in GC_2, then GC_2 will not not keep obj_2 alive because of that reference.
2. if your `ref_1->pointer.pointer` points to an object obj_1 in GC_1 then it will become a dangling reference when GC_1 is destroyed, and could cause a segmentation violation

**Comment by rptb1** [comment]
> how would we go about doing that?

Don't use more than one arena.  See https://github.com/Ravenbrook/mps/issues/259#issuecomment-1731576619

> at the moment yes

MPS finalization is the wrong tool.  See https://github.com/Ravenbrook/mps/issues/259#issuecomment-1731592474

**Comment by mgood7123** [comment]
> You are however welcome (and intended) to create multiple _pools_ within your arena. That might be what you need. Let's say you create two instances of the AMC pool class, amcA and amcB. Now, if an object in amc1 (objA) is keeping an object in amc2 (objB) alive, and you destroy pool amc1, then objB will (eventually) be recycled (and possibly finalized). If you don't want that to happen, you need a pointer to objB in an object that is still alive, or in a root. How you do that is up to you, but the simplest case is that you just do `objC.b = objB` where objC is some other object, which is presumably not dead because you're talking about it.

hmm in my use case, i am using multiple arena's to give isolated roots

eg

`arena_1` as X set of roots
`arena_2` has G set of roots

both may be on the same thread, or on different threads

seperate arena's are important to isolate the roots seen by one `GC pool` from the roots seen by another `GC pool`

**Comment by mgood7123** [comment]
> The reason I ask about external resources is that the MPS makes decisions about collection based on factors to do with space and time: how much memory is available, CPU time available, and so on. It is designed to manage memory in a way that the running program will not notice. It's _very good_ at keeping pause times down for interactive applications. It does not think about external resources or make plans based on them. If you're trying to clean up external resources (e.g. file handles) promptly you're probably using the wrong tool. Finalization is only an _optimization_ that the MPS offers _in passing_. You should not rely on finalization messages.

i dont intend to clean up "promptly" except when destroying a GC instance (to avoid leaking references)

**Comment by mgood7123** [comment]
> > true, however our goal is to `explicitly finalize all unfinalized objects`, eg `turn all black objects to white objects, and keep all white objects as white objects` such that `everything in the arena is now white`
> 
> See item 4 of finalization [Cautions](https://memory-pool-system.readthedocs.io/en/latest/topic/finalization.html#cautions):
> 
> > The only reliable way to ensure that all finalizable objects are finalized is to maintain a table of [weak references (1)](https://memory-pool-system.readthedocs.io/en/latest/glossary/w.html#term-weak-reference-1) to all such objects. The weak references don’t prevent the objects from being finalized, but you can iterate over the table at an appropriate point and finalize any remaining objects yourself.
> 
> We _might_ be able to give access to the list of remaining finalized objects, but we'd have to think hard about whether we can do this safely.

we dont care about the list of remaining finalized objects, we only care about objects which are still `live`

eg, if `r1` exists in a root, the GC will count `r1` as live, and not finalize nor collect it

we want the GC to see `r1` as `dead` even though it is `live`, this way we can forcefully finalize and collect it

if `r1` is attempted to be accessed in the future then it will be considered a user error for incorrectly destroying the GC (similar to incorrectly freeing memory which may still be referenced)

**Comment by rptb1** [comment]
> hmm in my use case, i am using multiple arena's to give isolated GC's

This is interesting.  Why do you think you need isolated GCs?  Garbage collection is an action on the global graph of all objects in a process.  (Except in very weird setups.)

If there's any chance at all that there are pointers between your different activities, they should share a single GC.  The GC acts on the global graph of objects.

> seperate arena's are important to isolate the roots/threads seen by one GC pool from the roots/threads seen by another GC pool

Why is that important to you?

> i dont intend to clean up "promptly" except when destroying a GC instance (to avoid leaking references)

That is what we mean by "promptly": at a time of your choosing.

> we dont care about the list of remaining finalized objects, we only care about objects which are still live

I'm sorry, I was not clear.  I meant the list of remaining objects *registered for finalization* that are *still alive*.

**Comment by mgood7123** [comment]
> > additionally, since all roots are destroyed, reference resurrection should become effectively impossible
> 
> That's not quite true in the MPS. When the MPS discovers a finalizable object, it add it to a [message queue in the arena](https://memory-pool-system.readthedocs.io/en/latest/topic/finalization.html#finalization-messages). The message queue is not GCd, but it is scanned, and so it is a bit like a root. So the message itself resurrects the object until it is deleted.

true, but under `normal` conditions we should not need to worry about such

eg all roots have been destroyed before collection, the finalizers cannot resserect themselves since they should not be able to access any internal state of our GC, the GC is correctly destroyed in cooperation with standard multi-thread safety

> In most scenarios where the MPS is part of a language run-time system, the program cannot run with all roots destroyed, because it cannot allocate any objects that won't die immediately!
> 
> However, if you're writing a program where only _some_ objects are managed by the MPS, it can make sense.
> 
> > the `only` exception to this, is a root `can be manually re-registered` in the arena somehow, which requires a reference gaining access to the arena it belongs to, and then creating a root and then storing itself inside that root
> 
> This is not necessary, because the finalization message keeps the object alive (see above).

**Comment by mgood7123** [comment]
> Aha! What you are calling `ref_1` is what we call an _object_, not a reference. What we could call a reference is the field obj->pointer.pointer.
> 
> You can copy objects wherever you like, but be warned:
> 
>     1. if your `ref_1->pointer.pointer` points to an object obj_2 in GC_2, then GC_2 will not not keep obj_2 alive because of that reference.
> 
>     2. if your `ref_1->pointer.pointer` points to an object obj_1 in GC_1 then it will become a dangling reference when GC_1 is destroyed, and could cause a segmentation violation

so `mps_reserve` (` + mps_commit`) reserves an `object`, not a reference

HOWEVER such object is `treated` like a reference, as it is looked for in roots to determine of the object is still reachable right?

**Comment by mgood7123** [comment]
> > hmm in my use case, i am using multiple arena's to give isolated GC's
> 
> This is interesting. Why do you think you need isolated GCs? Garbage collection is an action on the global graph of all objects in a process. (Except in very weird setups.)

to allow multiple collections to run in multiple threads at the same time (as long as a thread is not registered to more than 1 GC)

additionally to allow for lower collection times since parallel GC's would mean less work to do unless multiple GC's overlap threads

eg, if we had a seperate GC for each thread, and all threads never share data with eachother, then theoretically we would have fully parallel garbage collection, with `stop-the-world` never happening since all data is thread local

if we share data then we only incur `stop-the-world` for the threads we are sharing data with

granted we must be careful about thread registration since if a thread gets suspended during the registration process the `sigNext` might not be filled
```cpp
/*  ProtSetup -- global protection setup
 *
 *  Under Unix, the global setup involves installing a signal handler
 *  on PROT_SIGNAL to catch and handle page faults (see sigHandle).
 *  The previous handler is recorded so that it can be reached from
 *  sigHandle if it fails to handle the fault.
 *
 *  NOTE: There are problems with this approach:
 *    1. we can't honor the sa_flags for the previous handler,
 *    2. what if this thread is suspended just after calling signal(3)?
 *       The sigNext variable will never be initialized!
 */

void ProtSetup(void)
{
  struct sigaction sa;
  int result;

  sa.sa_sigaction = sigHandle;
  result = sigemptyset(&sa.sa_mask);
  AVER(result == 0);
  sa.sa_flags = SA_SIGINFO | SA_RESTART;

  result = sigaction(PROT_SIGNAL, &sa, &sigNext);
  AVER(result == 0);
}
```

> If there's any chance at all that there are pointers between your different activities, they should share a single GC. The GC acts on the global graph of objects.

yea, this is intended

> 
> > seperate arena's are important to isolate the roots/threads seen by one GC pool from the roots/threads seen by another GC pool
> 
> Why is that important to you?

to ensure parelization where possible

> > i dont intend to clean up "promptly" except when destroying a GC instance (to avoid leaking references)
> 
> That is what we mean by "promptly": at a time of your choosing.

alright, tho still, when a GC is destroyed, we MUST clean up as much as we possibly can to minimize memory leakage from both finalizers, live references, and unreclaimed memory

> > we dont care about the list of remaining finalized objects, we only care about objects which are still live
> 
> I'm sorry, I was not clear. I meant the list of remaining objects _registered for finalization_ that are _still alive_.

oh ok

**Comment by rptb1** [comment]
> HOWEVER such object is `treated` like a reference, as it is looked for in roots to determine if the object is still reachable right?

Please understand I'm using words pedantically so that we can understand each other.  It's a tricky subject!

An object allocated by reserve/commit in a [GC'd pool](https://memory-pool-system.readthedocs.io/en/latest/pool/intro.html#pool-class-properties) (such as one of class AMC) will be [scanned](https://memory-pool-system.readthedocs.io/en/latest/topic/scanning.html#scanning) to *look for* references inside it.  The *fields* in that object are treated as references when they are passed to e.g. MPS_FIX by your format scanner.

So the references are the fields that your [scan method](https://memory-pool-system.readthedocs.io/en/latest/topic/format.html#c.mps_fmt_scan_t) fixes.

Those references are *inside* the object.  We don't usually say that the object *is* a reference.

**Comment by rptb1** [comment]
> to ensure parelization where possible

The MPS GC is a coroutine of your program and does not care about threads.  You will not gain parallelization by creating arenas.

**Comment by mgood7123** [comment]
> > to ensure parelization where possible
> 
> The MPS GC is a coroutine of your program and does not care about threads. You will not gain parallelization by creating arenas.

yes but to actually perform a collection, all threads registered to such arena must be suspended for the duration of the collection, right?

**Comment by mgood7123** [comment]
> > HOWEVER such object is `treated` like a reference, as it is looked for in roots to determine if the object is still reachable right?
> 
> Please understand I'm using words pedantically so that we can understand each other. It's a tricky subject!
> 
> An object allocated by reserve/commit in a [GC'd pool](https://memory-pool-system.readthedocs.io/en/latest/pool/intro.html#pool-class-properties) (such as one of class AMC) will be [scanned](https://memory-pool-system.readthedocs.io/en/latest/topic/scanning.html#scanning) to _look for_ references inside it. The _fields_ in that object are treated as references when they are passed to e.g. MPS_FIX by your format scanner.
> 
> So the references are the fields that your [scan method](https://memory-pool-system.readthedocs.io/en/latest/topic/format.html#c.mps_fmt_scan_t) fixes.
> 
> Those references are _inside_ the object. We don't usually say that the object _is_ a reference.

alright :)

**Comment by mgood7123** [comment]
> > HOWEVER such object is `treated` like a reference, as it is looked for in roots to determine if the object is still reachable right?
> 
> Please understand I'm using words pedantically so that we can understand each other. It's a tricky subject!
> 
> An object allocated by reserve/commit in a [GC'd pool](https://memory-pool-system.readthedocs.io/en/latest/pool/intro.html#pool-class-properties) (such as one of class AMC) will be [scanned](https://memory-pool-system.readthedocs.io/en/latest/topic/scanning.html#scanning) to _look for_ references inside it. The _fields_ in that object are treated as references when they are passed to e.g. MPS_FIX by your format scanner.
> 
> So the references are the fields that your [scan method](https://memory-pool-system.readthedocs.io/en/latest/topic/format.html#c.mps_fmt_scan_t) fixes.
> 
> Those references are _inside_ the object. We don't usually say that the object _is_ a reference.

so in regards to references, `ref_1->pointer.pointer` itself is a `field` that we pass to `FIX` to tell it that that field may contain a reference, but `ref_1->pointer.pointer` itself `is not scanned` for references, eg `((FOO*)ref_1->pointer.pointer)->bar` cannot be scanned to see if `bar` is a reference or not because we do not know its type at compile-time, and we cannot make it a root since the object itself may be moved via a moving garbage collector, invalidating the registered root

**Comment by mgood7123** [comment]
assuming so, in regards to topology #260 and also actually keeping references alive in other references

we would need some way to add `objects` to our `fields` at runtime in an effiecent manner, while also binding the `fields` to other object `fields`, eg `a->field` to `b` and `a->ptr to b->ptr`

**Comment by rptb1** [comment]
> yes but to actually perform a collection, all threads registered to such arena (excluding the thead running the collection) must be suspended for the duration of the collection, right?

No, not quite.

During some parts of a collection, all threads that might read or write any reference (usually *all* threads in your process) must be suspended, so that we can atomically update the graph of all objects in the process.

It's not safe to have some threads registered to one arena, and some to another arena, if there is any chance that the threads will look at or update any object, or share references with each other.

This is because the GC must act globally on the entire state of the program.

The details are very subtle.  But your process and its threads are not stopped for the duration of the GC, only for short periods.  The MPS is [incremental](https://memory-pool-system.readthedocs.io/en/latest/glossary/i.html#term-incremental-garbage-collection) and at quite a fine resolution.  That is one reason it is successfully commercially: it does not interrupt the user even when there is a lot of GC to do.

**Comment by rptb1** [comment]
> assuming so, in regards to topology #260 and also actually keeping references alive in other references
> 
> we would need some way to add `objects` to our `fields` at runtime in an effiecent manner, while also binding the `fields` to other object `fields`, eg `a->field` to `b` and `a->ptr to b->ptr`, right?

I'm out of time for today but I hope to help you again soon.

**Comment by mgood7123** [comment]
> During some parts of a collection, all threads that might read or write any reference (usually _all_ threads in your process) must be suspended, so that we can atomically update the graph of all objects in the process.

would it be possible to limit such to only registered threads?

**Comment by mgood7123** [comment]
> > assuming so, in regards to topology #260 and also actually keeping references alive in other references
> > we would need some way to add `objects` to our `fields` at runtime in an effiecent manner, while also binding the `fields` to other object `fields`, eg `a->field` to `b` and `a->ptr to b->ptr`, right?
> 
> I'm out of time for today but I hope to help you again soon.

alright

**Comment by rptb1** [comment]
> would it be possible to limit such to only registered threads?

We must suspend all threads that can look at any object managed by the MPS.  Basically because we are going to move them around without the program noticing.

**Comment by mgood7123** [comment]
> > would it be possible to limit such to only registered threads?
> 
> We must suspend all threads that can look at any object managed by the MPS. Basically because we are going to move them around without the program noticing.

so i assume we must explicitly register threads because we have no api to automatically register all existing/created threads?

**Comment by rptb1** [comment]
The amount of time that the threads are suspended is kept short.  The MPS uses [barriers](https://memory-pool-system.readthedocs.io/en/latest/glossary/b.html#term-barrier-1) so that it can GC and move objects without the threads noticing, even while they are running.

**Comment by mgood7123** [comment]
hmm ok

**Comment by rptb1** [comment]
> so i assume we must explicitly register threads because we have no api to automatically register all existing/new threads?

We don't currently have any portable code to do this.  We could possibly add some convenience code by looking at how, e.g. the BDWGC finds threads.  It's much more "drop-in" than the MPS currently.

**Comment by mgood7123** [comment]
so

a GC registers threads as roots to scan

if a collect happens, ALL threads (both registered and unregistered) are `suspended` for short amounts of time

the more threads get registered, the more we need to scan

multiple pools can be used to seperate references, but `all references` in `all pools` must be scanned
likewise `all` registered thread stack/register roots must be scanned
likewise `all` `registered and unregistered` threads must be suspended for short amount during a GC

**Comment by mgood7123** [comment]
eg
```cpp
/* ThreadRingSuspend -- suspend all threads on a ring, except the
 * current one.
 */

static Bool threadSuspend(Thread thread)
{
  Res res;
  pthread_t self;
  self = pthread_self();
  if (pthread_equal(self, thread->id)) /* .thread.id */
    return TRUE;

  /* .error.suspend: if PThreadextSuspend fails, we assume the thread
   * has been terminated. */
  AVER(thread->context == NULL);
  res = PThreadextSuspend(&thread->thrextStruct, &thread->context);
  AVER(res == ResOK);
  AVER(thread->context != NULL);
  /* design.thread-manager.sol.thread.term.attempt */
  return res == ResOK;
}
```

**Comment by rptb1** [comment]
> a GC registers threads as roots to scan
> if a collect happens, ALL threads (both registered and unregistered) are suspended for short amounts of time

Not quite.

The client program (your code) [registers threads](https://memory-pool-system.readthedocs.io/en/latest/topic/thread.html) with the arena.  This allows the GC to suspend them.  It does not create roots.  The MPS does not automatically find threads.

Only registered threads are suspended.

However, if an unregistered thread accesses any managed objects in the arena or any references to those objects from roots, you could get heap corruption.  The corruption could occur much later and will be very difficult to debug.  So we would recommend you register all threads.  There are some circumstances (special debugging hacks usually) where you might want an unregistered thread.

Except in rare circumstances, you also want to create [thread roots](https://memory-pool-system.readthedocs.io/en/latest/topic/root.html#thread-roots) with your threads as well, so that they can freely reference the managed objects.

> multiple pools can be used to seperate references, but all references in all pools must be scanned

You can separate your objects into pools for many reasons, e.g. different management policy, different parameters, easy mass destruction, etc.

The MPS needs to *consider* all references in the whole world, but it does not scan them every time.  It uses [remembered sets](https://memory-pool-system.readthedocs.io/en/latest/glossary/r.html#term-remembered-set) to optimise the scanning.  A lot.

> likewise all registered thread stack/register roots must be scanned

All *thread roots* are scanned, but basically yes.

> likewise all registered and unregistered threads must be suspended for short amount during a GC collect

We don't suspend unregistered threads because we don't know about them.  But I think you mean thread *roots* here.

**Comment by rptb1** [comment]
> eg

You have found the relevant code :)

**Comment by mgood7123** [comment]
also
```
_`.sol.exclusive`: In order to meet `.req.exclusive`_, the arena
maintains a ring of threads (in ``arena->threadRing``) that have been
registered by the client program. When the MPS needs exclusive access
to memory, it suspends all the threads in the ring except for the
currently running thread. When the MPS no longer needs exclusive
access to memory, it resumes all threads in the ring.

_`.sol.exclusive.assumption`: This relies on the assumption that any
thread that might refer to, read from, or write to memory in
automatically managed pool classes is registered with the MPS. This is
documented in the manual under ``mps_thread_reg()``.
```

**Comment by mgood7123** [comment]
> The client program (your code) [registers threads](https://memory-pool-system.readthedocs.io/en/latest/topic/thread.html) with the arena. This allows the GC to suspend them. It does not create roots. The MPS does not automatically find threads.
> 
> Only registered threads are suspended.
> 
> However, if an unregistered thread accesses any managed objects in the arena or any references to those objects from roots, you could get heap corruption. The corruption could occur much later and will be very difficult to debug. So we would recommend you register all threads. There are some circumstances (special debugging hacks usually) where you might want an unregistered thread.

generally this would be the case

however if we can make certain assumptions about our managed objects, and assert that they hold true, the we can optimize for only select range of threads to be suspended (eg, ones that we can gaurentee `will not` touch our managed objects we do not require being suspended)

**Comment by mgood7123** [comment]
> > multiple pools can be used to seperate references, but all references in all pools must be scanned
> 
> You can separate your objects into pools for many reasons, e.g. different management policy, different parameters, easy mass destruction, etc.
> 
> The MPS needs to _consider_ all references in the whole world, but it does not scan them every time. It uses [remembered sets](https://memory-pool-system.readthedocs.io/en/latest/glossary/r.html#term-remembered-set) to optimise the scanning. A lot.

alright

> > likewise all registered thread stack/register roots must be scanned
> 
> All _thread roots_ are scanned, but basically yes.
> 
> > likewise all registered and unregistered threads must be suspended for short amount during a GC collect
> 
> We don't suspend unregistered threads because we don't know about them. But I think you mean thread _roots_ here.

so all registered threads must be suspended then (as previously stated, MPS cannot suspend a thread that it has not been registered with)

**Comment by mgood7123** [comment]
hmmm

i implemented a small scanner but my object does not seem to be collecting

```cpp
// the compiler is not allowed to optimize out functions that use volatile pointers
static inline void managed_object_volatile_memset(void * ptr, char value, size_t length_in_bytes) {
  printf("volatile memset\n");
  volatile char* s = (char*)(ptr);
  volatile char* e = s + length_in_bytes;
#ifdef __cplusplus
  std::fill(s, e, 0);
#else
  for(;s!=e;s++) *s = value;
#endif
}

#ifdef __cplusplus
#define MANAGED_OBJECT_ZERO_AND_DELETE(T, ptr) static_cast<T*>(ptr)->~T(); managed_object_volatile_memset(ptr, 0, sizeof(T)); free(ptr)
#else
#define MANAGED_OBJECT_ZERO_AND_DELETE(T, ptr) managed_object_volatile_memset(ptr, 0, sizeof(T)); free(ptr)
#endif

#define MANAGED_OBJECT_FIX(ref) \
  do { \
    mps_addr_t _addr = (ref); /* copy to local to avoid type pun */ \
    mps_res_t res = MPS_FIX12(ss, &_addr); \
    if (res != MPS_RES_OK) return res; \
    (ref) = _addr; \
  } while(0)
```

```cpp
TEST(MPS_finalize, 4) {
    ManagedObjState state;
    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);

    volatile managed_obj_t b = managed_obj_make_scanned_with_finalizer(
        &state, new B(),
        // scanner
        +[](mps_ss_t ss, void*p) -> mps_res_t {
            MPS_SCAN_BEGIN(ss) {
                MANAGED_OBJECT_FIX(static_cast<B*>(p)->a);
            } MPS_SCAN_END(ss);
            return MPS_RES_OK;
        },
        // finalizer
        +[](void*p){ MANAGED_OBJECT_ZERO_AND_DELETE(B, p); }
    );
    volatile managed_obj_t b_a = managed_obj_make_with_finalizer(&state, new B_A(), +[](void*p){ delete static_cast<B_A*>(p); });
    MANAGED_OBJECT_SCANNED_CAST(B, b)->a = b_a;

    b_a = nullptr;

    printf("first collect, b should not be collected\n");
    printf("first collect, b_a should not be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("first collect, b should not be collected\n");
    printf("second collect, b_a should not be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    b = nullptr;

    printf("third collect, b may be collected\n");
    printf("third collect, b_a may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("fourth collect, b may be collected\n");
    printf("fourth collect, b_a may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    managed_obj_deinit(&state);
}
```
````cpp
[ RUN      ] MPS_finalize.4
thread stack bottom: 0x7fff40817000
reserved and comitted object 0x7f217271a000 with pointer 0x1ed42b0
reserved and comitted object 0x7f217271a020 with pointer 0x1ed42d0
first collect, b should not be collected
first collect, b_a should not be collected
Total Collection duration: 0 seconds, 0 milliseconds, 528 microseconds
first collect, b should not be collected
second collect, b_a should not be collected
Total Collection duration: 0 seconds, 0 milliseconds, 340 microseconds
third collect, b may be collected
third collect, b_a may be collected
object 0x7f2172780000 with pointer 0x1ed42b0 is being finalized.
~B() (child is B_A)
volatile memset
Total Collection duration: 0 seconds, 0 milliseconds, 591 microseconds
fourth collect, b may be collected
fourth collect, b_a may be collected
Total Collection duration: 0 seconds, 0 milliseconds, 421 microseconds
deinit - destroy all roots
deinit - destroyed all roots
deinit - collect
object 0x7f2172780000 with pointer 0x1ed42d0 is being finalized.
~B_A() (parent is A)
Total Collection duration: 0 seconds, 0 milliseconds, 546 microseconds
deinit - collected
deinit - shutdown
[       OK ] MPS_finalize.4 (7 ms)
```

**Comment by mgood7123** [comment]
if i set `p1` to a field and then fix it i get the same as `b->a = p1`

```cpp
      case MANAGED_OBJECT_TYPE_SCANNED_POINTER: {
        MANAGED_OBJECT_FIX(obj->scanned_pointer.pointer);
```
```cpp
    volatile managed_obj_t p1 = managed_obj_make_with_finalizer(&state, new B_A(), +[](void*p){ delete static_cast<B_A*>(p); });
    volatile managed_obj_t p2 = managed_obj_make_scanned_with_finalizer(
        &state, p1, // B_A (p1) is finalized after all roots have been destroyed
        // new B(),
        // scanner
        +[](mps_ss_t ss, void*p) -> mps_res_t {
        //     MPS_SCAN_BEGIN(ss) {
        //         // printf("b is %p, a is %p, fix b->a\n", p, ((B*)p)->a);
        //         MANAGED_OBJECT_FIX(p);
        //     } MPS_SCAN_END(ss);
            return MPS_RES_OK;
        },
        // finalizer
        +[](void*p) {}
    );
```

**Comment by mgood7123** [comment]
hmm

if we assign it to a volatile or pass it to a function that cannot be inlined then the compiler seems to put the variable onto the register/stack frame even in -O0

similar to https://github.com/Ravenbrook/mps/issues/259#issuecomment-1729395559

> this seems like with optimizations, attempting to bring a variable into the local stack of a function MAY bind it to the frame, resulting in it being uncollectable until the frame ends

> this in turn CAN be troublesome if we attempt to init and destroy the GC in the same frame, since the GC cannot reclaim a variable that is bound to the same frame as it (or even may leak to a parent frame variable)
(leaking would technically be UB as it would yield a reference that exists but is untracked since the GC has been destroyed and the reference was still alive at such point)

```cpp
TEST(MPS_finalize, 2_null) {
    ManagedObjState state;
    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);

    volatile managed_obj_t p1 = managed_obj_make_with_finalizer(&state, new B_A(), +[](void*p){ delete static_cast<B_A*>(p); });
    volatile void * p3 = p1;

    p1 = nullptr;

    printf("first collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("second collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("third collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("forth collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("fifth collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    managed_obj_deinit(&state); // p1 is actually collected here
}
```

**Comment by mgood7123** [comment]
seems to be related to https://github.com/Ravenbrook/mps/issues/259#issuecomment-1729395559

**Comment by mgood7123** [comment]
```cpp
void level_2(ManagedObjState & state) {
    volatile managed_obj_t p1 = managed_obj_make_with_finalizer(&state, new B_A(), +[](void*p){ delete static_cast<B_A*>(p); });
    volatile managed_obj_t p2 = managed_obj_make_with_finalizer(&state, p1, +[](void*p) {} );

    managed_obj_assert(p2, MANAGED_OBJECT_TYPE_SCANNED_POINTER)->scanned_pointer.pointer = nullptr; // remove p1

    p1 = nullptr;
    p2 = nullptr;

    // p1 can be destroyed now, no references to it should exist

    printf("first collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("second collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("third collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p
}

void level_1(ManagedObjState & state) {
    level_2(state);
    printf("forth collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("fifth collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("sixth collect, p may be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p
}

TEST(MPS_finalize, 2_null) {
    ManagedObjState state;
    managed_obj_init(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE);

    level_1(state);

    // p1 should have been destroyed by now

    managed_obj_deinit(&state); // p1 is actually destroyed here
}
```

if we do this then our `p1` (`B_A)` gets destroyed in the `fifth` collection while `p2` gets destroyed in the `first` collection

**Comment by mgood7123** [comment]
like

```cpp
    volatile managed_obj_t p2 = managed_obj_make_with_finalizer(&state, p1, +[](void*p) {} );

    managed_obj_assert(p2, MANAGED_OBJECT_TYPE_SCANNED_POINTER)->scanned_pointer.pointer = nullptr; // remove p1
```
should be the same as
```cpp
    volatile managed_obj_t p2 = managed_obj_make_with_finalizer(&state, nullptr, +[](void*p) {} );
```

but for some reason it isnt

**Comment by mgood7123** [comment]
using stack frames seems to work, as even if the compiler attaches the variable to the frame/register they will be gone when the frame exits

```
l_1 < outside func(), p1 and p2 do not exist here, they cannot be found in stack/registers
  func();
    l_2 < inside func(), p1 and p2 exist here, in stack/registers
  ... < func() return, p1 and p2 do not exist here, they cannot be found in stack/registers
```

**Comment by mgood7123** [comment]
also using custom memory works
```cpp
TEST(MPS_finalize, 2_null_user_mem_6) {
    ManagedObjState state;

    managed_obj_t memory[mem_count];

    managed_obj_init_with_user_memory(&state, MANAGED_OBJECT_DEFAULT_ARENA_SIZE, memory, sizeof(void*)*mem_count);

    memory[0] = managed_obj_make_with_finalizer(&state, new B_A(), +[](void*p){ delete static_cast<B_A*>(p); });
    memory[1] = managed_obj_make_scanned_with_finalizer(
        &state, new B(),
        // scanner
        //
        // if we do not scan 'a' then 'memory[0]' will be collected
        //   when 'memory[0]' is set to 0x0
        //
        +[](mps_ss_t ss, void*p) -> mps_res_t {
            MPS_SCAN_BEGIN(ss) {
                MANAGED_OBJECT_FIX(static_cast<B*>(p)->a);
            } MPS_SCAN_END(ss);
            return MPS_RES_OK;
        },
        // finalizer
        +[](void*p){ MANAGED_OBJECT_ZERO_AND_DELETE(B, p); }
    );
    MANAGED_OBJECT_SCANNED_CAST(B, memory[1])->a = memory[0];

    printf("first collect, p should not be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("second collect, p should not be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("third collect, p should not be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    memory[0] = nullptr;

    printf("forth collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("fifth collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("sixth collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    memory[1] = nullptr;

    printf("seventh collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("eigth collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    printf("ninth collect, p should be collected\n");
    managed_obj_collect(&state); // should collect the original pointer assigned to p

    // p1 should have been destroyed by now

    managed_obj_deinit(&state); // p1 is actually destroyed here
}
```

**Comment by mgood7123** [comment]
imma close this for now
